src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/StringLiteralEqualityCheck.java

```diff
@@ -112,12 +112,35 @@
     @Override
     public void visitToken(DetailAST ast) {
         final boolean hasStringLiteralChild =
-            ast.findFirstToken(TokenTypes.STRING_LITERAL) != null
-                || ast.findFirstToken(TokenTypes.TEXT_BLOCK_LITERAL_BEGIN) != null;
+                ast.findFirstToken(TokenTypes.STRING_LITERAL) != null
+                    || ast.findFirstToken(TokenTypes.TEXT_BLOCK_LITERAL_BEGIN) != null
+                    || areStringsConcatenated(ast);
 
         if (hasStringLiteralChild) {
             log(ast, MSG_KEY, ast.getText());
         }
     }
 
+    /**
+     * Checks whether string literal or text block literals are concatenated.
+     *
+     * @param ast ast
+     * @return {@code true} if string literal or text block literals are concatenated
+     */
+    private static boolean areStringsConcatenated(DetailAST ast) {
+        DetailAST plusAst = ast.findFirstToken(TokenTypes.PLUS);
+        boolean result = false;
+        while (plusAst != null) {
```

Comment:
I think there is no need to check all nested plus tokens. you can check only top level token. In case any of its operands is string, type of other operand is irrelevant, string concatenation always results in string type. 
Therefore nested plus signs checks can be removed. What do you think?

Reply:
Initially, I also thought the same but doing so will miss cases like-
```
String s = "a";
String p = "b";
if (s != s + "p" + p) { // violation
}
```
Relevant AST-
```
|--EXPR -> EXPR
    `--NOT_EQUAL -> !=
        |--IDENT -> s
        `--PLUS -> +
            |--PLUS -> +
            |   |--IDENT -> s
            |   `--STRING_LITERAL -> "p"
            `--IDENT -> p
```
As top-level `PLUS` does not contain any `STRING_LITERAL`

Reply:
is there a violation now?
I see test input
```
if ((s += "asd") != s + "p" + p) { // violation
```
but it is not 100% clear what violation is produced by second part of expression, but not first one.
Can you please add your example to make it 100% clear?

Reply:
> is there a violation now? I see test input
> 
> ```
> if ((s += "asd") != s + "p" + p) { // violation
> ```
> 
> but it is not 100% clear what violation is produced by second part of expression, but not first one. Can you please add your example to make it 100% clear?

The violation is being produced by the second part i.e. `s + "p" + p` but yes the test case is not clear, added the one I mentioned above.